### PR TITLE
Fix rocksdb oom with g.V().hasLabel().count()

### LINKED ISSUE #793: gremlin计数 HugeGraph OOM
### Actual behavior 实际表现
在数据量比较大的情况下， 运行类似下面语句时，出现HugeGraph OOM

```
g.V().hasLabel('vbsku').count() 
g.V().hasLabel('vbsku').groupCount().by(properties('p_trade').value())
```

### Steps to reproduce the problem 复现步骤

运行上面命令时，查看rocksdb的统计信息，最高点如下:
image
过两分钟后，rocksdb的统计信息如下：
image

根据[Memory-usage-in-RocksDB](https://github.com/facebook/rocksdb/wiki/Memory-usage-in-RocksDB)和[RocksDBMetrics](https://github.com/hugegraph/hugegraph/blob/master/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBMetrics.java)， 当前hugegraph中rocksdb统计内存，只统计了其中的三个部分，Blocks pinned by iterators没有统计在内。
目前还不能确认是系统的哪个部分把内存吃掉了。

#### Vertex/Edge summary 数据量
1000w级别的点边数据

### Specifications of environment 环境信息
- hugegraph version: 0.10.4
- operating system: centos 6， 32CPUs， 180G RAM
- hugegraph backend: rocksdb

### PR DESCRIPTION
**Pull request by javeme**
fix #793
rebase for #798

### CODE REVIEW

**Comment by javeme** [comment]
TODO: let tx.queryXX() return CIter

**Comment by javeme** [comment]
TODO add gremlin job test with `g.V()` and `g.V().hasLabel(xx)`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinaryEntryIterator.java [hunk 1]
```diff
@@ -76,6 +76,9 @@ protected final boolean fetch() {
             } else if (merged == this.current) {
                 // The next entry belongs to the current entry
                 assert this.current != null;
+                if (this.sizeOf(this.current) >= INLINE_BATCH_SIZE) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinaryEntryIterator.java]: same as cassandra and hbase

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/IdHolder.java [hunk 2]
```diff
@@ -19,89 +19,220 @@
...
 
-    public boolean paging() {
-        return this.idsFetcher != null;
+    public static class PagingIdHolder extends IdHolder {
+
+        private final Function<ConditionQuery, PageIds> fetcher;
+
+        /**
+         * For paging situation
+         * @param query         original query
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/IdHolder.java]: 'original query' align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 3]
```diff
@@ -1211,6 +1275,10 @@ private static boolean hasNullableProp(HugeElement element, Id key) {
         return indexLabels;
     }
 
+    private static void increaseLimit(Query query) {
+        query.limit(query.limit() * 2L + 8L);
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: What's the basis?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 4]
```diff
@@ -1285,11 +1353,57 @@ public static IndexQueries of(IndexLabel il, ConditionQuery query) {
...
         }
+
+        public Query rootQuery() {
+            if (this.size() > 0) {
+                return this.values().iterator().next().rootOriginQuery();
+            }
+            return null;
+        }
+
+        public Query asQuery() {
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: asJointQuery maybe clearer

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/IdHolder.java [hunk 5]
```diff
@@ -19,89 +19,218 @@
...
+                throw new NoSuchElementException();
+            }
+            return this;
+        }
+
+        @Override
+        public PageIds fetchNext(String page, long batchSize) {
+            E.checkArgument(page == null,
+                            "Not support page parameter by BatchIdHolder");
+            E.checkArgument(batchSize >= 0L,
```
[the remark this answers is not in the document]
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/page/IdHolder.java]: yes, it's 0 when hit the last page

**Pull request merged**